subsys/net/ip/net_pkt.c

```diff
@@ -2176,6 +2188,472 @@ struct net_pkt *net_pkt_clone(struct net_pkt *pkt, s32_t timeout)
+	alloc_len = pkt_get_size(pkt) - net_pkt_get_len(pkt);
+	if (!alloc_len) {
+		/* In case nothing is there, it will account for header
+		 * space estimation
+		 */
+		hdr_len = pkt_headers_buffer_length(pkt,
+						    net_pkt_family(pkt),
+						    proto);
+	}
+
+	/* Calculate the maximum that can be allocated depending on size */
+	alloc_len = pkt_buffer_length(pkt, size + hdr_len, proto, alloc_len);
+
+	NET_DBG("Data allocation maximum size %zu (requested %zu)",
+		alloc_len, size);
+
+	if (timeout != K_NO_WAIT && timeout != K_FOREVER) {
+		u32_t diff = k_uptime_get_32() - alloc_start;
+
+		timeout -= min(timeout, diff);
+	}
+
+#if NET_LOG_LEVEL >= LOG_LEVEL_DBG
+	buf = pkt_alloc_buffer(pkt->slab == &tx_pkts ?
+			       &tx_bufs : &rx_bufs,
+			       alloc_len, timeout,
+			       caller, line);
+#else
+	buf = pkt_alloc_buffer(pkt->slab == &tx_pkts ?
+			       &tx_bufs : &rx_bufs,
+			       alloc_len, timeout);
+#endif
+
+	if (!buf) {
+		NET_ERR("Data buffer allocation failed.");
+		return -ENOMEM;
+	}
+
+	if (!pkt->buffer) {
+		pkt->buffer = buf;
+	} else {
+		net_buf_frag_insert(net_buf_frag_last(pkt->buffer), buf);
+	}
+
+	return 0;
+}
+
+#if NET_LOG_LEVEL >= LOG_LEVEL_DBG
+static struct net_pkt *pkt_alloc(struct k_mem_slab *slab, s32_t timeout,
+				 const char *caller, int line)
+#else
+static struct net_pkt *pkt_alloc(struct k_mem_slab *slab, s32_t timeout)
+#endif
+{
+	struct net_pkt *pkt;
+	int ret;
+
+	if (k_is_in_isr()) {
+		timeout = K_NO_WAIT;
+	}
+
+	ret = k_mem_slab_alloc(slab, (void **)&pkt, timeout);
+	if (ret) {
+		return NULL;
+	}
+
+	memset(pkt, 0, sizeof(struct net_pkt));
+
+	pkt->ref = 1;
+	pkt->slab = slab;
+
+#if NET_LOG_LEVEL >= LOG_LEVEL_DBG
+	net_pkt_alloc_add(pkt, true, caller, line);
+#endif
+
+	return pkt;
+}
+
+
+#if NET_LOG_LEVEL >= LOG_LEVEL_DBG
+struct net_pkt *net_pkt_alloc_debug(s32_t timeout,
+				    const char *caller, int line)
+#else
+struct net_pkt *net_pkt_alloc(s32_t timeout)
+#endif
+{
+#if NET_LOG_LEVEL >= LOG_LEVEL_DBG
+	return pkt_alloc(&tx_pkts, timeout, caller, line);
+#else
+	return pkt_alloc(&tx_pkts, timeout);
+#endif
+}
+
+#if NET_LOG_LEVEL >= LOG_LEVEL_DBG
+struct net_pkt *net_pkt_rx_alloc_debug(s32_t timeout,
+				       const char *caller, int line)
+#else
+struct net_pkt *net_pkt_rx_alloc(s32_t timeout)
+#endif
+{
+#if NET_LOG_LEVEL >= LOG_LEVEL_DBG
+	return pkt_alloc(&rx_pkts, timeout, caller, line);
+#else
+	return pkt_alloc(&rx_pkts, timeout);
+#endif
+}
+
+#if NET_LOG_LEVEL >= LOG_LEVEL_DBG
+static struct net_pkt *pkt_alloc_on_iface(struct k_mem_slab *slab,
+					  struct net_if *iface, s32_t timeout,
+					  const char *caller, int line)
+#else
+static struct net_pkt *pkt_alloc_on_iface(struct k_mem_slab *slab,
+					  struct net_if *iface, s32_t timeout)
+
+#endif
+{
+	struct net_pkt *pkt;
+
+#if NET_LOG_LEVEL >= LOG_LEVEL_DBG
+	pkt = pkt_alloc(slab, timeout, caller, line);
+#else
+	pkt = pkt_alloc(slab, timeout);
+#endif
+
+	if (pkt) {
+		net_pkt_set_iface(pkt, iface);
```

Comment:
So, the whole purpose of these 2 extra functions to just set ->iface member of a pkt? Well, that's rather non-orthogonal API, I wonder if it really comes from tangible improvement with such design.

Reply:
As I said in my earlier comment, I might just hide/remove net_pkt_alloc, net_pkt_alloc_on_iface. I thought it could be useful to somebody, I haven't found any usage for myself at least.